### PR TITLE
e2e test network

### PR DESCRIPTION
**Pull request by leonz789**
## Description

This PR supports testutil/network to work properly with exocorenetwork.
It starts a exocore network with n in process nodes with n validators. We can write UT to do e2e test by interact with this in-process network.
And this PR has a bug fix for NST balance change fail caused by Inappropriate assetID configure. 
- [x] lcoanetwork test framework
- [x] gentx, queryclient, ed25519 
- [x] add test cases
- [x] docs

## Notable Changes
- `testutil/netowrk/`: e2e frameworkd
- `tests/e2e/`: e2e test cases
  - `bank/`
  - `oracle/`: LST price update, NST balance change, slashing
- `x/oracle/`: oracle NST blanceupdate bug fix
----

Closes #XXX

## Release Notes

- **New Features**
  - Introduced new querying methods for various components of the Exocore network.
  - Added functionality for sending transactions to Ethereum precompile contracts.
  - Implemented end-to-end testing suites for the bank and oracle modules, validating core functionalities like coin transfers and price creation.

- **Improvements**
  - Enhanced error handling and logging for balance updates and slashing processes.
  - Streamlined price data handling by transitioning from `big.Int` to string types, simplifying the codebase.

- **Bug Fixes**
  - Improved clarity in error messages related to signature verification and public key validation during transaction simulations.

- **Chores**
  - Updated dependencies in the `go.mod` file for improved compatibility and functionality.
  - Removed outdated GitHub Actions workflows related to E2E testing.

These updates enhance user experience, improve system reliability, and ensure robust testing of critical functionalities.

### CODE REVIEW

File: testutil/network/genesis_data.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+			{
+				AssetBasicInfo: assetstypes.AssetInfo{
+					Name:             "NST ETH",
+					MetaInfo:         "native restaking ETH",
+					Address:          "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
+					LayerZeroChainID: 101,
+				},
+				StakingTotalAmount: sdkmath.NewInt(5000),
+			},
+		},
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/genesis_data.go]: updated

File: testutil/network/util.go [hunk 2]
```diff
@@ -26,8 +34,11 @@
 	genutiltypes "github.com/cosmos/cosmos-sdk/x/genutil/types"
 	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
 	govv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
 
+	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/evmos/evmos/v16/server"
+	evmostypes "github.com/evmos/evmos/v16/types"
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/util.go:39]: it's certainly used

File: testutil/network/util.go [hunk 3]
```diff
@@ -247,3 +275,169 @@
...
+			Operator:    operatorAccAddress.String(),
+			AssetsState: assetsStates,
+		})
+	}
+
+	DefaultGenStateAssets.Deposits = depositsByStakers
+	DefaultGenStateAssets.OperatorAssets = operatorsAssets
+
+	return DefaultGenStateAssets, nil
+}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/util.go]: expected

File: x/oracle/keeper/msg_server_update_params.go [hunk 4]
```diff
@@ -66,7 +66,7 @@ func (ms msgServer) UpdateParams(goCtx context.Context, msg *types.MsgUpdatePara
 	}
 	// set updated new params
 	ms.SetParams(ctx, p)
-	_ = GetAggregatorContext(ctx, ms.Keeper)
-	cs.AddCache(cache.ItemP(p))
+	_ = ms.Keeper.GetAggregatorContext(ctx)
+	ms.Keeper.GetCaches().AddCache(cache.ItemP(p))
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/msg_server_update_params.go:70]: the return value will never be nil.

File: x/oracle/keeper/params.go [hunk 5]
```diff
@@ -74,8 +74,8 @@
 			t.AssetID = strings.Join([]string{t.AssetID, oInfo.AssetID}, ",")
 			k.SetParams(ctx, p)
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/params.go:75]: duplicated IDs had been checked earlier

File: testutil/network/util.go [hunk 6]
```diff
@@ -247,3 +275,169 @@
...
+	totalDepositAmount := depositAmount.Mul(nInt)
+	depositsByStakers := make([]assetstypes.DepositsByStaker, 0, len(DefaultGenStateAssets.Tokens)*n)
+	operatorsAssets := make([]assetstypes.AssetsByOperator, 0, n)
+	nAssets := len(DefaultGenStateAssets.Tokens)
+	for i := 0; i < nAssets; i++ {
+		DefaultGenStateAssets.Tokens[i].StakingTotalAmount = totalDepositAmount
+	}
+	for _, operatorAccAddress := range operatorAccAddresses {
+		// use the same address []byte for operator(exo..) and staker(0x...), both derived from the same pubkey and since evmos use ethsecp256k1, this address is generated from keccak-256(.) instead of ripemd160(sha256(.))
+		stakerAddrStr := hexutil.Encode(operatorAccAddress)
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/util.go:307]: that's expected behavior

File: x/oracle/keeper/keeper.go [hunk 7]
```diff
@@ -58,6 +70,7 @@ func NewKeeper(
 		delegationKeeper: delegationKeeper,
 		assetsKeeper:     assetsKeeper,
 		authority:        authority,
+		memStore:         new(memoryStore),
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/keeper.go:75]: expected

File: x/oracle/module.go [hunk 8]
```diff
@@ -155,18 +155,18 @@
...
 	// init caches and aggregatorContext for node restart
 	// TODO: try better way to init caches and aggregatorContext than beginBlock
-	once.Do(func() {
-		_ = keeper.GetCaches()
-		_ = keeper.GetAggregatorContext(ctx, am.keeper)
-	})
+	//	once.Do(func() {
+	_ = am.keeper.GetCaches()
+	_ = am.keeper.GetAggregatorContext(ctx)
+	// })
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/module.go]: That won't be invoked concurrently by single node. And it's necessary for this test framework

File: x/oracle/keeper/params.go [hunk 9]
```diff
@@ -74,8 +74,8 @@ func (k Keeper) RegisterNewTokenAndSetTokenFeeder(ctx sdk.Context, oInfo *types.
...
 			if !ctx.IsCheckTx() {
-				_ = GetAggregatorContext(ctx, k)
-				cs.AddCache(cache.ItemP(p))
+				_ = k.GetAggregatorContext(ctx)
+				k.GetCaches().AddCache(cache.ItemP(p))
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/params.go:78]: That return value will not be `nil`

File: testutil/network/util.go [hunk 10]
```diff
@@ -247,3 +271,201 @@
...
+}
+
+// stakingAmount, each operator opt in evry AVS with stakingAmount of every assets
+// each avs suppport all assets
+// each operator opts in every avs
+// each operator deposited and self staked all assets with: (depsitAmount, stakingAmount)
+// initial price for every asset is 1 USD
+func NewGenStateOperator(operatorAccAddresses []sdktypes.AccAddress, consPubKeys []string, commissionRate sdkmath.LegacyDec, chainID string, optedAVSAddresses []string, stakingAmount sdkmath.Int, genStateAssets assetstypes.GenesisState) (operatortypes.GenesisState, error) {
+	// total stakingAmount one operator holds among all assets
+	stakingAmount = stakingAmount.Mul(sdkmath.NewInt(int64(len(genStateAssets.Tokens))))
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/util.go:356]: safe for test case as initialization

File: testutil/network/tx_internal.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+)
+
+func (n *Network) SendTxInternal(msgs []sdk.Msg, keyName string, kr keyring.Keyring) (*ctypes.ResultBroadcastTx, error) {
+	signedTxBytes, err := n.genSignedTxInternal(msgs, keyName, kr)
+	if err != nil {
+		return nil, err
+	}
+	// txClient.BroadcastTx()
+	return n.Validators[0].RPCClient.BroadcastTxSync(context.Background(), signedTxBytes)
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/tx_internal.go:20]: There must be at least one validator exists

File: testutil/network/util.go [hunk 12]
```diff
@@ -193,29 +169,85 @@ func initGenFiles(cfg Config, genAccounts []authtypes.GenesisAccount, genBalance
...
+	var feemarketGenState feemarkettypes.GenesisState
+	cfg.Codec.MustUnmarshalJSON(cfg.GenesisState[feemarkettypes.ModuleName], &feemarketGenState)
+	l := len(cfg.MinGasPrices) - len(cfg.NativeDenom)
+	minGasPrice := sdkmath.ZeroInt()
+	if l > 0 {
+		minGasPriceStr := cfg.MinGasPrices[:l]
+		minGasPrice, _ = sdkmath.NewIntFromString(minGasPriceStr)
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/util.go]: the 2nd result value is not error

File: testutil/network/network.go [hunk 13]
```diff
@@ -10,12 +10,13 @@ import (
...
 	"os"
 	"path/filepath"
+	"sort"
 	"strings"
 	"sync"
 	"testing"
 	"time"
 
-	"cosmossdk.io/math"
+	sdkmath "cosmossdk.io/math"
```
[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/network.go:19]: This is fixed in previous commit, why do you still report that ?

[the remark this answers is not in the document]
Comment (leonz789) [reply testutil/network/network.go:19]: @cloud8little

File: tests/e2e/oracle/create_price.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+	kr2 := s.network.Validators[2].ClientCtx.Keyring
+	creator2 := sdk.AccAddress(s.network.Validators[2].PubKey.Address())
+
+	//	kr3 := s.network.Validators[2].ClientCtx.Keyring
+	//	creator3 := sdk.AccAddress(s.network.Validators[2].PubKey.Address())
+
+	priceTest1R1 := price1.updateTimestamp()
+	priceTimeDetID1R1 := priceTest1R1.getPriceTimeDetID("9")
+	priceSource1R1 := oracletypes.PriceSource{
```
[the remark this answers is not in the document]
Comment (leonz789) [reply tests/e2e/oracle/create_price.go:69]: false alert

File: x/oracle/keeper/aggregator/aggregator.go [hunk 15]
```diff
@@ -18,26 +18,31 @@ type priceWithTimeAndRound struct {
...
 	power  *big.Int
 }
 
-func (r *reportPrice) aggregate() *big.Int {
-	if r.price != nil {
+func (r *reportPrice) aggregate() string {
+	if len(r.price) > 0 {
 		return r.price
 	}
 	tmp := make([]*big.Int, 0, len(r.prices))
```
Comment (bwhour) [inline x/oracle/keeper/aggregator/aggregator.go:31]: // Validate input ？
    if len(r.prices) == 0 {
        return "0"      // Or return error/default value
    }

File: x/oracle/keeper/native_token.go [hunk 16]
```diff
@@ -20,27 +20,28 @@ import (
 // undelegate: update operator's price, operator's totalAmount, operator's totalShare, staker's share
 // msg(refund or slash on beaconChain): update staker's price, operator's price
 
-type NSTETHAssetID string
+type NSTAssetID string
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go:24]: some spelling erros in this file.

1. "deopsit" should be "deposit" (in the UpdateNSTValidatorListForStaker)  
2. "shoule" should be "should" (in the UpdateNSTByBalanceChange)  
3. "ingore" should be "ignore" (in the UpdateNSTByBalanceChange)  
4. "lenft" should be "left" (in the parseBalanceChangeCapped)

**Review by TimmyExogenous — APPROVED** [review]

**Review by trestinlsd — APPROVED** [review]

File: x/oracle/keeper/aggregator/aggregator.go (hunk 15 above)
Comment (leonz789) [reply x/oracle/keeper/aggregator/aggregator.go:31]: no, empty string means no final price yet

File: x/oracle/keeper/aggregator/aggregator.go [hunk 17]
```diff
@@ -98,31 +103,26 @@ func (agg *aggregator) fillPrice(pSources []*types.PriceSource, validator string
...
-					price:     priceBigInt,
+					price:     pTmp.Price,
 					decimal:   pTmp.Decimal,
 					timestamp: pTmp.Timestamp,
-					//			detRoundId: p.DetId,
 				}
 				report.prices[pSource.SourceID] = pTR
 			} else {
-				pTR.price, _ = (&big.Int{}).SetString(pSource.Prices[0].Price, 10)
+				pTR.price = pSource.Prices[0].Price
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/aggregator/aggregator.go:113]: string of non is acceptable

File: x/oracle/keeper/aggregator/calculator.go [hunk 18]
```diff
@@ -10,34 +10,34 @@ import (
...
 func (r *roundPrices) updatePriceAndPower(pw *priceAndPower, totalPower *big.Int) (updated bool, confirmed bool) {
-	if r.price != nil {
+	if len(r.price) > 0 {
 		confirmed = true
 		return
 	}
 	for _, item := range r.prices {
-		if item.price.Cmp(pw.price) == 0 {
+		if item.price == pw.price {
 			item.power = new(big.Int).Add(item.power, pw.power)
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/aggregator/calculator.go:41]: That's fine. If price represent a number, it should be integer

File: x/oracle/keeper/aggregator/context.go [hunk 19]
```diff
@@ -179,13 +183,12 @@ func (agc *AggregatorContext) FillPrice(msg *types.MsgCreatePrice) (*PriceItemKV
...
 	if listFilled := feederWorker.do(msg); listFilled != nil {
-		// record this message for performance evaluation(used for slashing)
 		feederWorker.recordMessage(msg.Creator, msg.FeederID, listFilled)
-		if finalPrice := feederWorker.aggregate(); finalPrice != nil {
+		if finalPrice := feederWorker.aggregate(); len(finalPrice) > 0 {
 			agc.rounds[msg.FeederID].status = roundStatusClosed
 			feederWorker.seal()
 			return &PriceItemKV{agc.params.GetTokenFeeder(msg.FeederID).TokenID, types.PriceTimeRound{
-				Price:   finalPrice.String(),
+				Price:   finalPrice,
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/aggregator/context.go:191]: we accept nan string for NST senario

**Review by mikebraver — APPROVED** [review]

**Review by bwhour — APPROVED** [review]

**Review by cloud8little — APPROVED** [review]

**Pull request merged**